UCT/IB/FENCE: added resource check to fence

hoopoepg opened the pull request:
- added available resources check to IB fence call to allow to push this op into pending queue

Commit:
UCT/IB/FENCE: added resource check to fence

- added available resources check to IB fence call to
  allow to push this op into pending queue

brminich reviewed (commented):
i think we only need that for "strong" fence mode, which should be enabled when AR is detected (or enabled by config)

shamisp commented:
What is the difference between strong fence and regular fence ?

hoopoepg commented:
> What is the difference between strong fence and regular fence ?

in PRM there are several fence modes: read/atomic mode, strong mode (all outgoing operations must be completed), etc...

Commit:
Merge branch 'master' of https://github.com/openucx/ucx into topic/add-resource-check-into-ib-fence

Commit:
UCT/IB/FENCE: added resource check to fence

- added strong-ordering fence flag
- fence will not be applied in case if AR is not actual
- updated tests

hoopoepg commented:
> i think we only need that for "strong" fence mode, which should be enabled when AR is detected (or enabled by config)

ok, updated

shamisp commented:
Maybe just a comment to the code "all outgoing operations must be completed" to describe what it actually means.

Commit:
UCT/IB/FENCE: added resource check to fence

- added fence modes description

hoopoepg commented:
> Maybe just a comment to the code "all outgoing operations must be completed" to describe what it actually means.

ok, added descriptions

hoopoepg commented:
/azp run

hoopoepg commented:
@brminich more comments?

hoopoepg commented on src/uct/ib/base/ib_iface.h:
```diff
@@ -293,6 +293,7 @@ struct uct_ib_iface {
         uint8_t               max_inl_cqe[UCT_IB_DIR_NUM];
         uint8_t               port_num;
         uint8_t               sl;
+        uint8_t               sl_ar;
```
ok, renamed

hoopoepg commented on src/uct/ib/dc/dc_mlx5_ep.c:
```diff
@@ -258,6 +258,12 @@ ucs_status_t uct_dc_mlx5_ep_fence(uct_ep_h tl_ep, unsigned flags)
     uct_dc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
 
+    if (!iface->super.super.super.config.sl_ar) {
+        return uct_base_ep_fence(tl_ep, flags);
```
really strange suggestion, ok, done

brminich replied:
what is strange?

hoopoepg replied:
resource check depends from AR...

hoopoepg commented on src/uct/ib/rc/accel/rc_mlx5_ep.c:
```diff
@@ -535,7 +535,13 @@ ucs_status_t uct_rc_mlx5_ep_atomic_cswap32(uct_ep_h tl_ep, uint32_t compare, uin
+        return uct_base_ep_fence(tl_ep, flags);
+    }
+
+    UCT_RC_CHECK_RES(&iface->super, &ep->super);
```
fixed

hoopoepg commented on src/uct/ib/rc/verbs/rc_verbs_ep.c:
```diff
@@ -511,7 +511,14 @@ ucs_status_t uct_rc_verbs_ep_flush(uct_ep_h tl_ep, unsigned flags,
+    uct_rc_verbs_ep_t *ep       = ucs_derived_of(tl_ep, uct_rc_verbs_ep_t);
+
+    if (!iface->super.super.config.sl_ar) {
+        return uct_base_ep_fence(tl_ep, flags);
```
reimplemented via uct_rc_ep_fence

Commit:
UCT/IB/FENCE: added resource check to fence

- refactoring for resource check

hoopoepg commented:
@brminich more comments?

Commit:
UCT/IB/FENCE: added resource check to fence

- removed condition for resource check
- updated fence flag initialization

hoopoepg commented:
@brminich, reverted some changes

gleon99 commented:
Hi @hoopoepg - we're cleaning up the UCX PR backlog and plan to close this PR in 7 days. Please let us know if you'd like to keep it open.
